### ISSUE TITLE
Unable to create new access token from only refresh token.

### ISSUE THREAD

**Issue by duaneking** [aws/aws-sdk-net-extensions-cognito#69]
The Readme shows a refresh token scenario that requires that the users username be known in advance so that the users CognitoUser object can be retrieved and its methods thus called.

The code itself in the readme is not functional; the constants "username" etc are clearly placeholders meant to show that a developer should use the correct username for that specific CognitoUser in question, the same as I would presume the other fields are meant to be used for pool id and client id configuration parameters,, but that's not even represented in the signature of the method itself, so at best its a non-working example that's actively misleading, if not outright hostile to users, because it gives them false hope of a working solution.

And in case I was wrong on this, I actually tried to use the code after setting up config values.. and found that attempting to actually use it just results in an error in my tests, in my case a "NotAuthorizedException" in my tests, no doubt because that user doesn't exist in my test user pool as I'm otherwise able to generate a access token log in a user, etc.  It does not work, so I'm confused as to why the readme actively misdirects people.

I'm also noticing that it uses DateTime.Now instead of DateTime.UtcNow in a few places, so there is another issue.

Another issue: When an API client hits a 401 and the server challenges the bearer, it now is told that the AccessToken has expired, but the redirect doesn't pass in any other data at all.

Simply put:  How do we actually get a new auth token from only a refresh token that according to the spec is everything we should need, if that is all we have, in the call?  How can we securely create a new access tokens for a user without knowing the username of the user in advance, using only the refresh token itself as input?

 I'm asking for a working solution. Not hypotheticals. Thanks.

**Comment by ashishdhingra** [comment]
Hi @duaneking,

Please refer the below working code sample that had capability to use RefreshToken. Kindly note that this is a sample (console) application and you might want to move the secrets to a configuration file. And yes, **you should have the correct username** that exists in CognitoUserPool to use the RefreshToken. Kindly note that this sample uses `Amazon.Extensions.CognitoAuthentication` package.

```C#
using System;
using System.Threading.Tasks;
using Amazon;
using Amazon.CognitoIdentityProvider;
using Amazon.Extensions.CognitoAuthentication;
using Amazon.Runtime;

namespace CognitoStartWithRefreshTokenAuthAsync
{
    class Program
    {
        private static string userPoolId = "us-east-2_25NH0hy66";
        private static string clientId = "5lih3k6ejher88kksacjmoqijt";
        private static string clientSecret = "1ajhb49kub1go43e7t2qree9chsojg1g8ggds0k0lum387ahc352";
        private static RegionEndpoint regionEndpoint = RegionEndpoint.USEast2; // Change the region appropriately. Credentials are loaded from default profile.

        static void Main(string[] args)
        {
            Console.WriteLine("User Name: ");
            string userName = Console.ReadLine();
            while (string.IsNullOrWhiteSpace(userName))
            {
                Console.WriteLine("Please enter a valid User Name.");
                userName = Console.ReadLine();
            }

            Console.WriteLine("Do you have a Refresh Token (Y/N): ");
            char hasRefreshTokenResponse = Convert.ToChar(Console.Read());
            bool hasRefreshToken = (char.ToLower(hasRefreshTokenResponse) == 'y');
            Console.WriteLine();

            Console.WriteLine("Password: ");
            string password = Console.ReadLine();
            while (string.IsNullOrWhiteSpace(password) && !hasRefreshToken)
            {
                Console.WriteLine("Please enter a valid Password.");
                password = Console.ReadLine();
            }

            Console.WriteLine("Existing Refresh Token: ");
            string refreshToken = Console.ReadLine();
            while (string.IsNullOrWhiteSpace(refreshToken) && hasRefreshToken)
            {
                Console.WriteLine("Please enter a valid Refresh Token.");
                refreshToken = Console.ReadLine();
            }

            AuthFlowResponse authFlowResponse = (!string.IsNullOrWhiteSpace(refreshToken) ? GetCredsFromRefreshAsync(userName, refreshToken).GetAwaiter().GetResult() : GetCredentials(userName, password).GetAwaiter().GetResult());
        }

        public static async Task<AuthFlowResponse> GetCredentials(string userName, string password)
        {
            var provider = new AmazonCognitoIdentityProviderClient(new AnonymousAWSCredentials(), FallbackRegionFactory.GetRegionEndpoint());
            var userPool = new CognitoUserPool(userPoolId, clientId, provider, clientSecret);
            var user = new CognitoUser(userName, clientId, userPool, provider, clientSecret);

            AuthFlowResponse authResponse = await user.StartWithSrpAuthAsync(new InitiateSrpAuthRequest()
            {
                Password = password
            }).ConfigureAwait(false);

            while (authResponse.AuthenticationResult == null)
            {
                if (authResponse.ChallengeName == ChallengeNameType.NEW_PASSWORD_REQUIRED)
                {
                    Console.WriteLine("Enter your desired new password: ");
                    string newPassword = Console.ReadLine();

                    authResponse = await user.RespondToNewPasswordRequiredAsync(new RespondToNewPasswordRequiredRequest()
                    {
                        SessionID = authResponse.SessionID,
                        NewPassword = newPassword
                    });
                }
                else if (authResponse.ChallengeName == ChallengeNameType.SMS_MFA)
                {
                    Console.WriteLine("Enter the MFA Code sent to your device: ");
                    string mfaCode = Console.ReadLine();

                    authResponse = await user.RespondToSmsMfaAuthAsync(new RespondToSmsMfaRequest()
                    {
                        SessionID = authResponse.SessionID,
                        MfaCode = mfaCode

                    }).ConfigureAwait(false);
                }
                else
                {
                    Console.WriteLine("Unrecognized authentication challenge.");
                    return null;
                }
            }

            return authResponse;
        }

        public static async Task<AuthFlowResponse> GetCredsFromRefreshAsync(string userName, string refreshToken)
        {
            AmazonCognitoIdentityProviderClient provider = new AmazonCognitoIdentityProviderClient(new AnonymousAWSCredentials(), FallbackRegionFactory.GetRegionEndpoint());
            CognitoUserPool userPool = new CognitoUserPool(userPoolId, clientId, provider, clientSecret);

            CognitoUser user = new CognitoUser(userName, clientId, userPool, provider, clientSecret);

            user.SessionTokens = new CognitoUserSession(null, null, refreshToken, DateTime.Now, DateTime.Now.AddHours(1));

            InitiateRefreshTokenAuthRequest refreshRequest = new InitiateRefreshTokenAuthRequest()
            {
                AuthFlowType = AuthFlowType.REFRESH_TOKEN_AUTH
            };

            return await user.StartWithRefreshTokenAuthAsync(refreshRequest).ConfigureAwait(false);
        }
    }
}
```

Hope this helps.

Thanks,
Ashish

**Comment by duaneking** [comment]
Thank you for posting code that proves my reported defect actually exists.

Now how would we go about using only a refresh token of a previously logged in user without the username to generate a new access token for that user, per the spec this currently violates?

**Comment by ashishdhingra** [comment]
> Thank you for posting code that proves my reported defect actually exists.
> 
> Now how would we go about using only a refresh token of a previously logged in user without the username to generate a new access token for that user, per the spec this currently violates?

Hi @duaneking,

The `Amazon.Extensions.CognitoAuthentication` package provides wrapper for using classes in `Amazon.CognitoIdentityProvider` and `Amazon.CognitoIdentity` namespaces. It's design makes actions (like using the RefreshToken, etc.) on top of the `CognitoUser`, refer https://github.com/aws/aws-sdk-net-extensions-cognito/blob/e9a2447ba179b7795eede4619906f48d0d3ee863/src/Amazon.Extensions.CognitoAuthentication/CognitoUserAuthentication.cs#L33.

So, unfortunately, you cannot use RefreshToken without the CognitoUser when using this package. Although you may try using the `Amazon.CognitoIdentityProvider` directly, but that would add complexity for your use case.

Thanks,
Ashish

**Comment by duaneking** [comment]
The problem is that requiring the username actively violates security standards and opens up users to more risk by leaking identity data that the refresh token itself is designed to protect against.

A refresh token, once expired, is designed to be opaque and not leak identity information.  Requiring the username in this way violates that, and so such an implementation would fail a DFIR audit as it would require the identity linked to the refresh token be leaked in the requests/repositories that store it.

This design is thus inherently insecure, and violates the claim that Amazon Cognito is HIPAA eligible and PCI DSS, SOC, and ISO/IEC 27001, ISO/IEC 27017, ISO/IEC 27018, and ISO 9001 compliant because according to you, this is designed to require additional credentials and that violates https://tools.ietf.org/html/rfc6749#section-6, do I understand you correctly?

**Comment by ashishdhingra** [comment]
> The problem is that requiring the username actively violates security standards and opens up users to more risk by leaking identity data that the refresh token itself is designed to protect against.
> 
> A refresh token, once expired, is designed to be opaque and not leak identity information. Requiring the username in this way violates that, and so such an implementation would fail a DFIR audit as it would require the identity linked to the refresh token be leaked in the requests/repositories that store it.
> 
> This design is thus inherently insecure, and violates the claim that Amazon Cognito is HIPAA eligible and PCI DSS, SOC, and ISO/IEC 27001, ISO/IEC 27017, ISO/IEC 27018, and ISO 9001 compliant because according to you, this is designed to require additional credentials and that violates https://tools.ietf.org/html/rfc6749#section-6, do I understand you correctly?

Hi @duaneking,

Thanks for your response. As mentioned earlier, this package is designed to invoke requests mostly around CognitoUser. I'm not sure if design could be changed. 

I also see that you are trying to automate some tests which require you to store this data, may be in the code itself (in turn in the code repository). However storing this data in repository itself is a bad design, and this should instead be stored in external data source (e.g. database) and could be fetched at runtime.

In the actual production application as well, this data would be stored in some store, and may follow the below workflow:
- User logs in.
- Application looks up the identity database for the tokens.
- If the token is expired, it tries to use the RefreshToken to authenticate.
- Once authenticated, it stores the new token in database.

This is just the general thought on the application workflow.

However, I will have a developer look at your concern and would post any updates accordingly.

Thanks,
Ashish

**Comment by duaneking** [comment]
> I also see that you are trying to automate some tests..

Not true. I have no idea where you got this bad assumption from.

The goal is actually to never store the tokens in any way on the server side, because if they are not stored then they can not be leaked.

**Comment by ashishdhingra** [comment]
> The goal is actually to never store the tokens in any way on the server side, because if they are not stored then they can not be leaked.

@duaneking Could you please elaborate on your use case on how you plan to store and use RefreshToken? Is the RefreshToken stored only in user session and used only when the access token expires (in this case also, RefreshToken is tied to a user and when user session ends, it gets rid of the tokens as well)?

**Comment by duaneking** [comment]
Also, your code above uses DateTime.Now and that's a defect, you should be using DateTime,.UtcNow in all cases.

**Comment by ashishdhingra** [comment]
> Ok so attempting to leverage that code or rewrite it but use the same call stack is resulting in nothing but exceptions.
> 

@duaneking Kindly check if you have configured the AWS credentials, refer [Credential and profile resolution](https://docs.aws.amazon.com/sdk-for-net/latest/developer-guide/creds-assign.html) for more details.

**Comment by duaneking** [comment]
With respect, everything is correctly configured as this same code base registers users, logs them in, generates access, id, and refresh tokens, ad also validates JWT's etc.  Refresh is the only thing that does not work.

This minimal code repos the exception in visual studio:

`
var refreshToken = "put the refresh token here";
var email = "user inside cognitopool yes replace this";
CognitoUser user = _userManager.FindByEmailAsync(email)?.Result;
user.SessionTokens = new CognitoUserSession(null, null, refreshToken, DateTime.UtcNow, DateTime.UtcNow.AddHours(1));
var refreshRequest = new InitiateRefreshTokenAuthRequest() { AuthFlowType = AuthFlowType.REFRESH_TOKEN_AUTH };
var authResponse = user.StartWithRefreshTokenAuthAsync(refreshRequest).ConfigureAwait(false).GetAwaiter().GetResult();
`

**Comment by ashishdhingra** [comment]
> With respect, everything is correctly configured as this same code base registers users, logs them in, generates access, id, and refresh tokens, ad also validates JWT's etc. Refresh is the only thing that does not work.
> 
> This minimal code repos the exception in visual studio:
> 
> `
> var refreshToken = "put the refresh token here";
> 
> var email = "user inside cognitopool yes replace this";
> 
> CognitoUser user = _userManager.FindByEmailAsync(email)?.Result;
> 
> user.SessionTokens = new CognitoUserSession(null, null, refreshToken, DateTime.UtcNow, DateTime.UtcNow.AddHours(1));
> 
> var refreshRequest = new InitiateRefreshTokenAuthRequest() { AuthFlowType = AuthFlowType.REFRESH_TOKEN_AUTH };
> 
> var authResponse = user.StartWithRefreshTokenAuthAsync(refreshRequest).ConfigureAwait(false).GetAwaiter().GetResult();
> `

@duaneking Thanks for your reply. I will give it a try using the sample app provided in the repo, adding the RefreshToken logic manually and let you know the findings.

**Comment by duaneking** [comment]
![image](https://user-images.githubusercontent.com/569341/114095805-bc6ebe00-9872-11eb-984a-d2182ba730c1.png)

Code to reproduce based on your posted code:

```
private void ACogitoGeneratedRefreshTokenIsNotValid(string clientId, string userPoolId, string email, string refreshToken)
{
var provider = new AmazonCognitoIdentityProviderClient(new AnonymousAWSCredentials(), FallbackRegionFactory.GetRegionEndpoint());
var userPool = new CognitoUserPool(userPoolId, clientId, provider);
CognitoUser user = new CognitoUser(email, clientId, userPool, provider);
user.SessionTokens = new CognitoUserSession(null, null, refreshToken, DateTime.UtcNow, DateTime.UtcNow.AddHours(1));
var refreshRequest = new InitiateRefreshTokenAuthRequest() { AuthFlowType = AuthFlowType.REFRESH_TOKEN_AUTH };

// This call throws every time.
var authResponse = user.StartWithRefreshTokenAuthAsync(refreshRequest).ConfigureAwait(false).GetAwaiter().GetResult();

Console.WriteLine(authResponse.AuthenticationResult.AccessToken);
}

```

**Comment by ashishdhingra** [comment]
> StartWithRefreshTokenAuthAsync

Hi @duaneking,

I was able to correctly execute the RefreshToken auth flow using Visual Studio:
<img width="1679" alt="Screen Shot 2021-04-08 at 4 53 36 PM" src="https://user-images.githubusercontent.com/67916761/114110020-635f5400-988b-11eb-914c-ef08b08038ff.png">

In case you app client has client secret as well, you would need to pass them to `CognitoUserPool` and `CognitoUser` constructors.

Hope this helps.

Thanks,
Ashish

**Comment by normj** [comment]
Hi @duaneking 

I agree the design of the CognitoUser and refreshing tokens is too dependent on knowing the user id. I tested and was able to use the `GetCredsFromRefreshAsync` code sample with passing null in for the user-id. I think we should work on making a more intuitive workflow for refreshing tokens then this user-id based approach. I am assuming you created your Cognito app client without generating a client secret. If you do create the app client with a client secret then you will have to have a user-id to generate the secret hash for the request.

Not sure why you are getting the invalid refresh token error. In my testing I got this error when I was accidently using a refresh token that was created from a different app client id then the app client I was using to refresh with. I have also seen issues when Device tracking is enabled but the device key is not passed along.

Thanks for pointing out the improper use of `DateTime.Now`. I updated the README.md file to use `DateTime.UtcNow.

Norm

**Comment by duaneking** [comment]
@ashishdhingra  The code I have still does not work.  I'm getting nothing but "refresh token is not valid" errors. This matter is not resolved.

@normj  This still repo's when the single user pool in that region only has only one client, so its not a wrong client issue either. 
 Device tracking has been turned off this entire time.  You are correct in that there is no secret for this test, and so per your feedback, a user id should not be required. .
 
Attempting to use null results in the same error:
![image](https://user-images.githubusercontent.com/569341/114324832-4a4fe080-9ae1-11eb-8676-8d15d52d61e1.png)


I have made 0 progress on this over the last few days and the AWS support we pay for has been absolutely worthless.

**Comment by normj** [comment]
Glad you have been able to make some headway and sorry for the frustrating experience getting to a working stage. Cognito does not currently update refresh tokens when using them to get new access tokens. So if you are getting new access tokens from the refresh call then you are working correctly.

Refresh tokens do expire. Their length is configurable when creating an app client for your Cognito user pool.

![image](https://user-images.githubusercontent.com/1653751/114627644-7054bc80-9c6a-11eb-84f0-29e48bf6ea48.png)

As whether refresh tokens should change after using them to get refresh tokens that is a question for the Cognito service. I'm on the AWS .NET Team and can just answer for this library. From my understanding through refresh tokens have their own lifecycle that is separate then access tokens and are used very differently. The access token is sent to the resource server and the refresh token is sent to the authorization server (Cognito) along with the client id and optional hashes of client secret if that was enabled. So far the refresh token to be compromised and useful the client id and client secret also have to be compromised. 

Again I'm not on the Cognito service and they would know better then I would. I believe you are also have conversations with them so I will focus this issue on things we can control with this library.

**Comment by HaikuJock** [comment]
@normj 
> 
> Thanks for pointing out the improper use of `DateTime.Now`. I updated the README.md file to use `DateTime.UtcNow`.
> 
> Norm

If we are using `DateTime.UtcNow` for the creation of the `CognitoUserSession` then shouldn't [CognitoUserSession.IsValid](https://github.com/aws/aws-sdk-net-extensions-cognito/blob/014a6c855f2576764b113c08dc39a99f4031a96c/src/Amazon.Extensions.CognitoAuthentication/CognitoUserSession.cs#L76) also use `DateTime.UtcNow`?

**Comment by duaneking** [comment]
@normj Refresh tokens should change after they are used even once as they should be expiring after use for security. Anything less means that an old key is now an attack vector.

**Comment by normj** [comment]
@HaikuJock Version 2.2.1 was just released which switches all usages of `DateTime.Now` to `DateTime.UtcNow`.

@duaneking Since refresh tokens creation is controlled by the service there is nothing this library can do for your request of generating a different refresh token every time it is used to get a new access token. I'm going to close this issue because I think we have answered the specific issue as best we can from the viewpoint of this library. I want to avoid this issue being used for other issues like the `DateTime` issue, which was a good catch, because it makes the issues hard to track. If you discover other issues with the library please open up new issues.

**Comment by duaneking** [comment]
There needs to be a bug filed against the cognito service for refresh tokens not changing after they are redeemed the first time; Expiring them out based only on time is a critical defect; And them using the SDK team as a meatshield for their service defects is not good engineering, it also actively promotes a lack of customer focus by that team.

The security expectation is that once used or redeemed to generate a new access token, a refresh token should change and that refresh token should no longer be able to generate a new access token, because a refresh token should only be able to generate a new access token and refresh token pair only once, and when used, should be considered used and expired.  Its a one time token, meant to be used once.
